presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java

```diff
@@ -83,9 +88,11 @@ public FileSingleStreamSpillerFactory(
             BlockEncodingSerde blockEncodingSerde,
             SpillerStats spillerStats,
             List<Path> spillPaths,
-            double maxUsedSpaceThreshold)
+            double maxUsedSpaceThreshold,
+            boolean spillCompressionEnabled,
+            boolean spillEncryptionEnabled)
     {
-        this.serdeFactory = new PagesSerdeFactory(requireNonNull(blockEncodingSerde, "blockEncodingSerde is null"), false);
+        this.serdeFactory = new PagesSerdeFactory(requireNonNull(blockEncodingSerde, "blockEncodingSerde is null"), spillCompressionEnabled);
```

Comment:
Did you consider applying compression over whole data stream instead of on page-by-page basis?
This would make the code simpler (e.g. no need for page markers), but at the same time you'd lose the ability to adapt to case when compression didn't yield a gain. Do you have such real-life cases?
I'd like to understand the motivation behind this choice.

@dain  WDYT?

Reply:
I didn't wan't to radically rewrite spill-to-disk along the way of implementing encryption, so that decision to do page-by-page compression was not re-evaluated for this PR.

I considered doing encryption over the whole stream and avoid having to change anything about the way that pages & markers work, but unfortunately pages are written to the file via opening the spill file in `APPEND` mode. Pages delivered over multiple calls now have a problem, because cipher's need to know when the end of stream occurs (last block padding) and holding the file open indefinitely to work around this problem (presumably until the first read occurs, at which point the stream could flush, close, and reads could begin) creates other undesirable changes in behavior (lots of open file descriptors for one, extending the lifespan of the output buffer for another).

I'm totally open to ideas about how to avoid those pitfalls and isolate compression or encryption to only exist within the file spiller though.

Reply:
I had the same thought on first reading, but after thinking more, I'm not sure there is a choice here.  I like the idea of just having encrypted output/input streams, and then not dealing with this at the higher level.  The down side is that this assumes that the spill system is stream oriented.  We may want to spill to a remote system by spraying pages to a bunch of stores, and in that case we wouldn't necessarily have output/input streams.  Of course, we may want to write pages to a store that has native encryption so this would be wasteful double encryption.  Also there is the problem of how we would configure this for multiple spill stores....  So.  My conclusion was "this seems like a good start and when we deal with the next store we can revisit this".